src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -102,13 +111,228 @@ private static async Task<VersionStamp> GetDocumentSemanticVersionAsync(Document
+            {
+                var tokens = await ComputeRequiredRange(computeStart: false, cachedTokens);
+                if (tokens is null)
+                {
+                    return null;
+                }
+
+                return tokens;
+            }
+            // 3b) The beginning of the range needs to be computed (i.e. the end is already cached).
+            else if (bestMatch.Range.End >= range.End)
+            {
+                var tokens = await ComputeRequiredRange(computeStart: true, cachedTokens);
+                if (tokens is null)
+                {
+                    return null;
+                }
+
+                return tokens;
+            }
+            else
+            {
+                // We shouldn't ever reach this point, so if we do we'll log an error and return.
+                logger.LogError("Caching error in DefaultRazorSemanticTokensInfoService.");
+                return null;
+            }
+
+            async Task<SemanticTokens?> ComputeRequiredRange(bool computeStart, SemanticTokens cachedTokens)
+            {
+                // We want to reduce the requested range to the part we don't already have cached.
+                var modifiedRange = computeStart
+                    ? new Range(range.Start, bestMatch.Range.Start)
+                    : new Range(bestMatch.Range.End, range.End);
+
+                var (partialTokens, isCSharpFinalized) = await GetSemanticTokensAsync(
+                    textDocumentIdentifier, documentSnapshot, documentVersion, modifiedRange, cancellationToken);
+                if (partialTokens is null)
+                {
+                    return null;
+                }
+
+                // Now we want to combine the ranges. It's possible for the two ranges to have
+                // duplicate tokens, so we can't just combine them automatically. We need to
+                // remove the overlapping tokens and in doing so may need to make minor adjustments
+                // to positioning.
+
+                var startTokens = computeStart ? partialTokens.Data.ToArray() : cachedTokens.Data.ToArray();
+                var endTokens = computeStart ? cachedTokens.Data.ToArray() : partialTokens.Data.ToArray();
+
+                // Compute the absolute end line/character of the starting set of tokens. This helps
+                // us later determine token duplicates.
+                var startLine = 0;
+                var startCharacter = 0;
+                for (var i = 0; i < startTokens.Length; i += 5)
+                {
+                    if (startTokens[i] != 0)
+                    {
+                        startCharacter = 0;
+                    }
+
+                    startLine += startTokens[i];
+                    startCharacter += startTokens[i + 1];
+                }
+
+                // Compute the absolute beginning line/character of the ending tokens.
+                // This loop also determines which tokens are duplicates.
```

Comment:
Hmmm just double checking, does this approach of always taking the "start tokens" when there's duplicates work on token boundaries? Aka I'm imagining if we're re-computing end-tokens and those are "more accurate" will skipping those end-tokens result in wrong or partial gaps in the colorization?